numba/np/arraymath.py

```diff
@@ -3877,6 +3877,22 @@ def impl(a, dtype=None):
 
         def impl(a, dtype=None):
             return np.array(a, ty)
+    elif isinstance(a, types.containers.ListType):
+        dt = a.dtype
+        if is_nonelike(dtype):
+            def impl(a, dtype=None):
+                l = len(a)
+                ret = np.empty(l, dtype=dt)
+                for i, v in enumerate(a):
+                    ret[i] = v
+                return ret
+        else:
```

Comment:
I don't see any difference between these two implementations, do we really need the distinction between none-typed and non-none-typed?

Reply:
I think we do need the distinction because you could convert an `int64` type list into a `float64` type list or vice versa, in which case you would specify a `dtype`.

Copy-pastable code to illustrate my point -
```python
import numpy as np
from numba import njit
from numba.typed import List


@njit
def convert_to_float(x):
    return np.asarray(x, dtype=np.float64)


a = List()
a.append(5)
a.append(3)
print(convert_to_float(a))
```

Reply:
Oh I see, yeah, that makes sense, so if a target `dtype` is specified, we need to use that.

In that case, how about the following slightly more concise variant (untested):

```python
target_dtype = a.dtype if is_nonelike(dtype) else dtype
def impl(a, dtype=None):
    l = len(a)
    ret = np.empty(l, dtype=target_dtype)
    for i, v in enumerate(a):
        ret[i] = v
    return ret
return impl
```

So the original and new types are known during typing time, so we can place the target dtype into a variable during typing and then just access it in the implementation?

Reply:
This works! 😄